staging/src/k8s.io/client-go/testing/fixture.go

```diff
@@ -416,7 +417,8 @@ func (t *tracker) add(gvr schema.GroupVersionResource, obj runtime.Object, ns st
 	t.objects[gvr][namespacedName] = obj
 
 	for _, w := range t.getWatches(gvr, ns) {
-		w.Add(obj)
+		// To avoid the object from being accidentally modified by watcher
+		w.Add(obj.DeepCopyObject())
```

Comment:
probably also needed in Delete when calling w.Delete(obj)

Reply:
We store obj in `t.objects` and do `w.Delete(obj)`, obj is deleted from `t.objects` map

https://github.com/kubernetes/kubernetes/blob/3f86955da2e55ad6e7fe025cdc705b94e03a0357/staging/src/k8s.io/client-go/testing/fixture.go#L453-L461

watcher receives obj that is not stored in `t.objects` map.

and I also test following code with `go test -race .`：
```go
	go func() {
		err := o.Create(testResource, testObj, ns)
		if err != nil {
			t.Errorf("test resource creation failed: %v", err)
		}
		_, err = o.Get(testResource, ns, testObj.GetName())
		// Delete obj
		err = o.Delete(testResource, ns, testObj.GetName())
		if err != nil {
			t.Errorf("test resource getting failed: %v", err)
		}

	}()
	out := <-watch.ResultChan()
	// Modify received object
	// Potential data race
	outCasted := out.Object.(*unstructured.Unstructured)
	outCasted.Object["data"] = strconv.Itoa(rand.Int())
```
So I think there is no data race :), What did I miss?

Reply:
if there are multiple watchers, and they modify the object in their delete handler, they should not observe each other's modifications

Reply:
Thank you, i get it and fix it. PTAL